### DISCUSSION TITLE
[RFC] 009 - 基于 YJS 的跨端同步

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#368] [RFCs]
## 背景

用户存在跨端同步的诉求。譬如使用 WebDAV 实现跨端同步
- #330
- #409
- #331 
- #491 
- #892 
- #911
- #971 
- #1202 
- #1184 

## 调研与讨论

目前来看，单纯使用 WebDAV 技术实现跨端同步，会出现很多关于同步一致性的问题:
- https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/3028
- https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/3001
- https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/2917
- https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/2892
- https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/2877
- https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/2837

原因在于同步时必然存在的数据冲突。而从根本上解决该问题，需要引入 [yjs](https://yjs.dev/) 这样的  CRDT 解决方案。

市面上暂时没有同时结合 Yjs 与 WebDAV 的技术方案，需要自行组装实现。 refs: https://sharegpt.com/c/AlFzX7s

大致思路上，是将现有的 persist 层基础上，添加一个 yjs 作为操作同步层。同时利用 YJS 实现数据在 indexedDB 层的持久化，与在 WebDAV 层的持久化。在这种实现架构下，后续甚至还可以通过接入 WebRTC 层，实现含有 AI 的多人群聊会话模式。

--- 

现在存在另外一个问题，是 LobeChat 现有数据结构的可扩展性问题。当前的存储结构并没有考虑到消息无限膨胀后的扩展情况。因此当一个 session 中存在大量消息时，有很大概率导致页面卡死。

为考虑代码的可扩展性，我们需要将现有的存储模式，从「本地化」单一大对象的形态，变成标准的「关系型」数据结构。同时实现按需加载以提升性能。

因此跨端同步的实施前提是在整个数据存储架构上做一次全面的重构，将现在单纯的 JSON 大对象模式转换为标准的关系型数据库结构。这对于实现本 RFC ，以及后续的全局搜索（ https://github.com/lobehub/lobe-chat/issues/334#issuecomment-1766302352 ）、服务端存储 #331  等需求都是重要的基石。

关于此部分重构，在 https://github.com/lobehub/lobe-chat/discussions/378 跟进

## 实现思路

YJS 的工作原理核心实现原理如下： （[refs](https://discuss.yjs.dev/t/could-anyone-help-me-to-visualize-how-yjs-works/1608/2)）

![image](https://github.com/lobehub/lobe-chat/assets/28616219/4450c7b1-f958-4f4c-898a-934b7eaa2422)

关于自定义 Provider 的实现可以参考：
- [y-localstorage](https://github.com/rozek/y-localstorage/blob/main/src/y-localstorage.ts)
- [y-mongodb-provider](https://github.com/MaxNoetzold/y-mongodb-provider)

**Comment by arvinxx** [comment]
## YJS 集成 Demo

将 YJS 的实例化过程独立于 store ，构建相应的双向绑定关系。再通过 provider 将状态同步到远程

代码如下：

```ts
const ydoc = new Doc();
const persistence = new IndexeddbPersistence('yjs-demo', ydoc);
const provider = new WebrtcProvider('yjs-demo', ydoc);

const ymap = ydoc.getMap();

useSessionStore.subscribe(
  (s) => s.activeId,
  (t) => {
    ymap.set('activeId', t);
  },
);

ymap.observe(() => {
  const id = ymap.get('activeId') as string;
  useSessionStore.getState().activeSession(id);
});

persistence.on('synced', () => {
  console.log('inited');
});
```

效果如下：

https://github.com/lobehub/lobe-chat/assets/28616219/315aeb04-07d5-4bf7-beb8-e2a128696ebd

----

上述为演示 demo ，在工程化的过程中还需要考虑到和组件的集成、SoT 选择、怎么处理水合逻辑等。

**Comment by Asuka109** [comment]
One disadvantage of yjs is that it is difficult to authenticate clients and impose access restrictions since it provides a peer-to-peer service, which may limit the implementation of some advanced features.

Some alternative solutions to yjs:

https://localfirstweb.dev/
https://twitter.com/vikingmute/status/1716655870459969791
https://twitter.com/waylybaye/status/1717038531833737344

**Reply by arvinxx** [reply]
> One disadvantage of yjs is that it is difficult to authenticate clients and impose access restrictions since it provides a peer-to-peer service, which may limit the implementation of some advanced features.

I think it seems not a problem, because yjs is not design for authentication, application can make the  authentication and then use yjs to sync data. 

For example, if we use WebRTC Provider of YJS, we need to create a room channel for the webrtc connection. And we can add a password  for the room. By using room name and the password ,we can hash them to a room channel id and use this id to create the channel.

and I have also discussed this with chatgpt, fyi: https://sharegpt.com/c/5aM8XJs

----

btw, this blog  [Local-first software](https://inkandswitch.com/local-first) is really amazing, I have read it latest year and then knew local-first and yjs. I even think that **local-first software is the next-stage web app**.

**Reply by Asuka109** [reply]
@arvinxx Because of the peer-to-peer design of yjs, any authenticated clients can access and modify any parts of the entire document. I am concerned that this may make it difficult to implement features such as collaborative editing and plugins that restrict read / write to specified data (if lobe chat wants to do that). Since Yjs doesn't care about the provider, we can control permissions in the provider. However, it is still difficult to restrict permissions for specific nodes in the document. We can also take it a step further and isolate content with different permissions in different documents and bind each to a provider. But before that, maybe we can first learn whether there are other out-of-the-box solutions available if the problem really matters.

**Comment by Asuka109** [comment]
btw, what is SoT?

**Reply by arvinxx** [reply]
SoT: Source of Truth.

We need to choose the SoT of data persist in local. Because the data store we want to refactor to is not same with yjs persist.

The yjs IndexedDB Provider is storage data to unit8Array like below. We can't use this data structure to implement the database search or similar features.
<img width="960" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/1c9f102c-041c-4b7b-bed8-3280d3d746e4">


We want to save data to IndexedDB more like relational dabase:
<img width="908" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/9f7ed5a6-c0a5-4377-9972-ed43f0ddff79">


So we need to decide how to solve the difference between two ways of saving data. This is also a challenge of engineering.

**Reply by Asuka109** [reply]
Didn't know that before. Can I refer to the description in [this article](https://tecky.io/zh_Hant/blog/%E5%B9%B3%E5%B8%B8%E4%BA%BA%E9%83%BD%E8%83%BD%E6%8E%8C%E6%8F%A1%E7%9A%84Programming%20%E5%8E%9F%E5%89%87/)?

**Reply by arvinxx** [reply]
yeah, you can

**Comment by arvinxx** [comment]
在 #378 实现基础上，会发现之前需要考虑的一些问题自然消除了:
1. 由于 store 不再使用 persist 中间件，因此无需再考虑水合问题。
2. 已经使用dexie 构建了标准的关系型数据库，并实现了基本CRUD与高级的检索能力，dexie所在table是本地 SoT
3. service 层将屏蔽 yjs/本地数据库/服务端请求之间的关系，在 store 层只需通过 `sessionService.getSessions` 这样的方法即可获取数据，无需关注数据来源。

如何结合YJS实现跨端同步，一个可能的思路如下：

----

实现Dexie.js和Yjs之间的映射涉及以下步骤：

1. **定义共享数据模型：**
   确定哪些数据需要在多个客户端之间同步。在Yjs中，您可以创建一个共享类型（如`Y.Map`或`Y.Array`），它将用作同步的基础。

2. **创建Yjs实例和文档：**
   创建一个Yjs文档(`Y.Doc`)作为数据的容器，并在其中添加共享类型实例。

3. **初始化Dexie.js数据库：**
   使用Dexie.js定义您的数据模型和索引，创建数据库实例，并准备用于CRUD操作的API。

4. **加载和同步初始状态：**
   当应用程序启动时，从Dexie.js数据库加载现有数据，并将其同步到Yjs文档中的相应共享类型。这可能涉及到将Dexie.js记录转换为Yjs数据结构的格式。

5. **监听Dexie.js的变化：**
   使用Dexie.js的Observable插件或其他方式监听数据库的变化，并将这些变化应用到Yjs的共享类型中。同样，当Yjs的共享类型发生变化时，也应该更新Dexie.js数据库。

6. **实现冲突解决：**
   每当发生数据变化时，使用Yjs的CRDT算法来解决可能出现的冲突。

下面是一个简化的示例，展示了如何实现Dexie.js和Yjs之间的基本映射：

```javascript
// 引入所需库
import Dexie from 'dexie';
import * as Y from 'yjs';
import { IndexeddbPersistence } from 'y-indexeddb';

// 定义Dexie数据库
const db = new Dexie('myDatabase');
db.version(1).stores({
  items: '++id,name'
});

// 创建Yjs文档
const ydoc = new Y.Doc();

// 创建Yjs类型用于同步
const yitems = ydoc.getMap('items');

// Yjs IndexedDB持久化
const persistence = new IndexeddbPersistence('myDatabase', ydoc);

// 将Dexie数据库的数据加载到Yjs
async function loadDataFromDexieToYjs() {
  const items = await db.items.toArray();
  items.forEach(item => {
    yitems.set(item.id.toString(), item);
  });
}

// 监听Dexie数据库的变化
db.items.hook('creating', (primaryKey, obj, transaction) => {
  yitems.set(primaryKey.toString(), obj);
});

db.items.hook('updating', (mods, primaryKey, obj, transaction) => {
  const newYitem = { ...obj, ...mods };
  yitems.set(primaryKey.toString(), newYitem);
});

db.items.hook('deleting', (primaryKey, obj, transaction) => {
  yitems.delete(primaryKey.toString());
});

// 监听Yjs的变化并更新Dexie数据库
yitems.observe(event => {
  event.changes.keys.forEach((change, key) => {
    const id = parseInt(key);
    if (change.action === 'add' || change.action === 'update') {
      const item = yitems.get(key);
      db.items.put({ ...item, id }); // Dexie需要id字段
    } else if (change.action === 'delete') {
      db.items.delete(id);
    }
  });
});

// 初始化加载数据
loadDataFromDexieToYjs();

// 示例：添加新项到Dexie.js（这也会更新Yjs）
db.items.put({ name: 'New Item' });

// 示例：更新Yjs（这也会更新Dexie.js）
yitems.set('1', { id: 1, name: 'Updated Item' });
```

在这个示例代码中，我们创建了一个Dexie数据库和一个Yjs文档，它们共享一个名为`items`的数据集合。我们定义了监听器来观察Dexie数据库和Yjs共享类型的变化，并且在变化发生的时候更新对方。这样，无论变化是在Dexie数据库发生还是在Yjs共享类型发生，另一方都会同步更新。

请注意，这只是一个基础示例，实际应用中可能需要更复杂的映射逻辑和错误处理。您需要根据自己的应用需求调整代码。

**Comment by arvinxx** [comment]
### 同步文件存储结构和策略的技术方案

#### 存储结构

在WebDAV服务器上，为AI会话应用设计的文件存储结构如下：

```
/{conversation_id}/
    /Snapshots/
        snapshot_{timestamp}.yjs  // 定期创建的完整快照
    /Deltas/
        delta_{from_version}_{to_version}.yjs  // 增量更新文件
    /Metadata/
        metadata.json  // 存储元数据，如参与者信息、创建时间等
    conversation_metadata.json  // 对话级别的元数据，如版本向量、最新状态等
```

每个用户和会话都有自己的目录。在每个会话目录中，包含三个子目录用于存储快照、增量更新（Deltas），以及元数据。

#### 同步策略

1. **定期快照策略**:
   - 在非高峰时段（例如每天午夜），自动创建每个会话的完整快照。这提供了一个基线状态，用于数据恢复和历史回溯。

2. **增量更新策略**:
   - 实时跟踪本地数据库中的变更。每5分钟，客户端应用收集自上次成功同步以来的所有变更，并打包成一个增量更新文件发送到服务器。

3. **版本控制**:
   - 使用Yjs内置的版本管理机制来跟踪和管理增量更新。每个增量更新包含所需的版本信息，以便服务端正确地应用这些更新。

4. **冲突解决**:
   - 服务端应用增量更新时，使用Yjs的CRDT机制解决任何潜在的冲突，确保最终数据的一致性。

5. **同步确认**:
   - 服务端成功应用增量更新后，向客户端发送确认消息。客户端在收到确认后更新本地的同步状态。

6. **离线支持**:
   - 在客户端离线时，变更被保存在本地。一旦连接恢复，立即同步这些变更。

7. **安全性和加密**:
   - 所有通过WebDAV传输的数据都使用SSL/TLS加密，确保数据在传输过程中的安全。

8. **存储优化和清理**:
   - 定期评估快照和增量更新文件的重要性，删除冗余或过时的文件，合并连续的增量更新以节省空间。
   - 保留关键的快照，例如每周第一次的快照或手动触发的快照。

9. **监控和日志记录**:
   - 对同步过程进行监控和日志记录，以便于问题排查和系统审计。

10. **用户界面**:
   - 提供用户界面允许用户手动触发快照创建，并管理自己的历史数据。

#### 技术和测试

- 选择支持Yjs库的前端框架和后端技术栈。
- 实现WebDAV客户端和服务器端逻辑。
- 对同步机制进行充分的测试，包括性能测试和冲突解决的测试。
- 根据反馈和测试结果优化同步策略和存储结构。

这个方案综合了之前讨论的存储结构和同步机制，提供了一个全面的技术路线图，旨在实现AI会话应用的高效和可靠的数据同步。

**Reply by Evsio0n** [reply]
Hi,這個存儲架構會考慮放在s3上嗎

**Comment by Dragode** [comment]
请问这个有计划的发布时间吗？谢谢
我很喜欢这个应用，希望能推给更多的人，感觉这个特性会是比较重要的一个卡点。

**Reply by arvinxx** [reply]
Q1

**Comment by boarpvu997** [comment]
配置存储、冲突解决应该是后端比较擅长的，是否考虑在后端实现这些逻辑？ @arvinxx

**Reply by arvinxx** [reply]
不考虑。LobeChat 没有计划提供一个后端和数据库。所以还是纯前端解决

**Comment by gaojunyang666** [comment]
作者大大开发的肿么样啦

**Reply by arvinxx** [reply]
最近在做 #737 。同步需求估计会在2月下旬开始，预计3月份应该能做完

**Comment by JinhuW** [comment]
是否考虑使用一个json file 来管理记录，并且利用Github的Gist 管理这个文件。把Github Gist 看成数据库

**Reply by arvinxx** [reply]
JsonFile 不太现实，原因是后面数据量一大，同步会越来越慢。

举个例子，我现在本地数据近3000条消息，几百个topics，导出 json 已经要几M了。这个量级的同步，用json效率会很低

**Comment by CXwudi** [comment]
> 我们需要将现有的存储模式，从「本地化」单一大对象的形态，变成标准的「关系型」数据结构。

如果能变成标准的「关系型」数据结构，那么是否可以仅仅依靠数据库的ACID特性（MySQL，PostgreSQL等均ACID）来保持同步。那这样的话用户是否可以只需填入一个URL（比如说`postgresql://${POSTGRESQL_USERNAME}:${POSTGRESQL_PASSWORD}@${POSTGRESQL_URL}:5432/${POSTGRESQL_DATABASE}`，无论在UI里填入还是在Docker environment variables里填入）即可开启同步

**Reply by arvinxx** [reply]
是的，你说的没错

**Comment by arvinxx** [comment]
## Roadmap 

1. 在 https://github.com/lobehub/lobe-chat/pull/1525 中实现 YJS 基础框架，作为实验性特性以征求社区反馈；
2. 根据社区实际用户反馈，决定是否进一步补充 Redis、WebDAV、S3 等 Provider 的接入；
